### PR TITLE
Mq2

### PR DESCRIPTION
**Pull request by sveinbjornt**
* Utility functions to generate Icelandic-language description strings of distances, time periods, króna sums
* Pi queries + more in arithmetic query module
* Lots and lots of improvements in most of the query modules, esp. date module
* Many additions to special module based on empirical review of logged queries
* New petrol station query module
* Set source of query answer
* Lots of new tests for query modules
* Refactoring
* OpenWeatherMap additions in weather module are incomplete, but not being invoked in the logic and can be safely merged. This PR is large enough as it is.

### CODE REVIEW

File: queries/petrol.py [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+    | "hvar" "er" "bensín" "ódýrast"
+    | "hvar" "er" "bensínið" "ódýrast"
+    | "hvar" "er" "bensínlítrinn" "ódýrastur"
+    | "hvar" "fæ" "ég" "ódýrasta" "bensínið"
+    | "hvar" "fæ" "ég" "ódýrasta" "bensínlítrann"
+    | "hvar" "er" "ódýrast" "að" "fylla" "á"? "tankinn"
+
+QPetrolClosestCheapestStation →
+    "ódýrt" "bensín" QPetrolNearMe?
+    | "hvar" "fæ" "ég" "ódýrt" "bensín" QPetrolNearMe?
```
Comment (vthorsteinsson) [inline queries/petrol.py]: Why is QPetrolNearMe optional? If left out, shouldn't this be a QPetrolCheapestStation?

File: queries/petrol.py [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+    | "hvar" "er" "ódýrast" "að" "fylla" "á"? "tankinn" QPetrolNearMe
+    | "hvar" "er" "ódýrt" "að" "fylla" "á"? "tankinn" QPetrolNearMe?
+    | "hvar" "get" "ég" "fyllt" "á" "tankinn" "ódýrt" QPetrolNearMe?
+
+QPetrolClosest →
+    "næsta" | "nálægasta"
+
+QPetrolNearMe →
+    QPetrolHere? QPetrolAround
+    | QPetrolAround
```
Comment (vthorsteinsson) [inline queries/petrol.py]: Delete second production, it is redundant and ambiguous

File: queries/petrol.py [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+def QPetrolCheapestStation(node, params, result):
+    result.qkey = "CheapestStation"
+
+
+def QPetrolClosestCheapestStation(node, params, result):
+    result.qkey = "ClosestCheapestStation"
+
+
+_PETROL_API = "https://apis.is/petrol"
+_PETROL_CACHE_TTL = 3600  # seconds
```
Comment (vthorsteinsson) [inline queries/petrol.py]: `# seconds, i.e. time-to-live of 1 hour`

File: queries/petrol.py [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+def _closest_petrol_station(loc):
+    """ Find petrol station closest to the given location. """
+    stations = _stations_with_distance(loc)
+    if not stations:
+        return None
+
+    # Sort by distance
+    dist_sorted = sorted(stations, key=lambda s: s["distance"])
+    if dist_sorted:
+        return dist_sorted[0]
```
Comment (vthorsteinsson) [inline queries/petrol.py]: Better to add explicit `return None`

File: queries/petrol.py [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+def _cheapest_petrol_station(loc):
+    stations = _stations_with_distance(loc)
+    if not stations:
+        return None
+
+    # Sort by price
+    price_sorted = sorted(stations, key=lambda s: s["bensin95"])
+    if price_sorted:
+        return price_sorted[0]
+
```
Comment (vthorsteinsson) [inline queries/petrol.py]: Better to add explicit `return None`

File: queries/petrol.py [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+        return None
+
+    # Filter out all stations that are not close by
+    filtered = filter(lambda x: x["distance"] <= _CLOSE_DISTANCE, stations)
+
+    # Sort by price
+    price_sorted = sorted(filtered, key=lambda s: s["bensin95"])
+    if price_sorted:
+        return price_sorted[0]
+
```
Comment (vthorsteinsson) [inline queries/petrol.py]: Better to add explicit `return None`

File: queries/petrol.py [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+
+def sentence(state, result):
+    """ Called when sentence processing is complete """
+    q = state["query"]
+    if "qtype" in result and "qkey" in result:
+        # Successfully matched a query type
+        try:
+            loc = q.location
+            if loc:
```
Comment (vthorsteinsson) [inline queries/petrol.py]: Location is actually not needed for QCheapestStation, so this is too strict

File: queries/petrol.py [hunk 8]
```diff
@@ -0,0 +1,297 @@
...
+        return None
+
+    # Sort by distance
+    dist_sorted = sorted(stations, key=lambda s: s["distance"])
+    if dist_sorted:
+        return dist_sorted[0]
+
+
+def _cheapest_petrol_station(loc):
+    stations = _stations_with_distance(loc)
```
Comment (vthorsteinsson) [inline queries/petrol.py]: Location is actually not needed for this query

File: queries/special.py [hunk 9]
```diff
@@ -66,6 +67,7 @@ def _capabilities(qs, q):
     "Af hverju eru Hafnfirðingar alltaf með stól úti á svölum? Svo sólin geti sest.",
     "Af hverju læðast Hafnfirðingar alltaf fram hjá apótekum? Til að vekja ekki svefnpillurnar.",
     "Af hverju fara Hafnfirðingar alltaf niður í fjöru um jólin? Til þess að bíða eftir jólabókaflóðinu.",
+    "Af hverju setti Hafnfirðingurinn skóna sína í fyrstinn? Hann vildi eignast kuldaskó.",
```
Comment (vthorsteinsson) [inline queries/special.py]: Typo: fyrstinn -> frystinn

File: queries/weather.py [hunk 10]
```diff
@@ -59,24 +62,52 @@
...
+    "ofsaveður",
+    "logn",
+    "lygn",
+    "blautur",
+    "bleyta",
+    "rigna",
+    "regnhlíf",
+    "votur",
+    "kaldur",
+    # "heitur", # Clashes with "hvað heitir X" etc.
```
Comment (vthorsteinsson) [inline queries/weather.py]: Hmm? That should not be the case for lemmas...

File: queries/weather.py [hunk 11]
```diff
@@ -222,6 +254,62 @@ def help_text(lemma):
...
+                "Could not read OpenWeatherMap API key from {0}".format(_OWM_KEY_PATH)
+            )
+            _OWM_API_KEY = ""
+    return _OWM_API_KEY
+
+
+def _postprocess_owm_data(d):
+    """ Restructure data from OWM API so it matches that provided by
+        the iceweather module. """
+    if not d:
```
Comment (vthorsteinsson) [inline queries/weather.py]: Still in development?

File: queries/wiki.py [hunk 12]
```diff
@@ -139,7 +139,16 @@ def _clean_answer(answer):
...
+    a = re.sub(r"\s,\s", ", ", a)
+    a = re.sub(r"\s\.\s", ". ", a)
+    # E.g. "100-700" becomes "100 til 700"
+    a = re.sub(r"(\d+)\-(\d+)", r"\1 til \2", a)
+    return a
+
+
+def _clean_voice_answer(answer):
+    a = answer.replace(" m.a. ", " meðal annars ")
+    a = a.replace(" þ.e. ", " það er ")
```
Comment (vthorsteinsson) [inline queries/wiki.py]: o.s.frv.? t.d.?

File: queries/distance.py [hunk 13]
```diff
@@ -53,24 +53,26 @@
     r"^hvað er ég langt í burtu frá (.+)$",
     r"^hversu langt er ég frá (.+)$",
     r"^hve langt er ég frá (.+)$",
-    r"^hvað er langt á (.+)$",
+    r"^hvað er langt\s?(?:austur|vestur|norður|suður)? á (.+)$",
```
[the remark this answers is not in the document]
Comment (sveinbjornt) [reply queries/distance.py]: This marks it as a non-capture group and it isn't included in matches.

File: queries/weather.py (hunk 11 above)
Comment (sveinbjornt) [reply queries/weather.py]: Yup. Will finish in next PR.

File: queries/date.py [hunk 14]
```diff
@@ -404,9 +498,81 @@ def QDateBankHoliday(node, params, result):
...
+
+
+def QDateSecondChristmasDay(node, params, result):
+    result["desc"] = "annar í jólum"
+    result["target"] = dnext(datetime(year=datetime.today().year, month=12, day=26))
+
+
+def QDateFirstDayOfWinter(node, params, result):
+    result["desc"] = "fyrsti vetrardagur"
+    result["target"] = None  # To be completed
```
[the remark this answers is not in the document]
Comment (sveinbjornt) [reply queries/date.py]: In the next PR :)

File: queries/petrol.py (hunk 1 above)
Comment (sveinbjornt) [reply queries/petrol.py]: A person in Akureyri asking "Hvar fæ ég ódýrt bensín?" is unlikely to want an answer telling him to go to Costco in Reykjavík some 500km away :).

**Pull request merged**